### PR TITLE
[Diagnostics] Clarify diagnostic for failed type conversion in subscript assignment

### PR DESCRIPTION
**Pull request by owenv**
Addresses [SR-6340](https://bugs.swift.org/browse/SR-6340)

Currently,

```swift
var d: [Int: String] = [:]
d[0]=0
```

results in the error,
```
error: cannot assign value of type 'Int' to type 'String?'
```

This PR changes the error message for subscript assignment so it now reads,
```
error: cannot assign value of type 'Int' to subscript of type 'String?
```

It's kind of unfortunate the new, subscript-specific messages don't use the existing assignment diagnostic definitions, but it keeps the `diagnoseContextualConversionError` additions much simpler

### CODE REVIEW

File: include/swift/AST/DiagnosticsSema.def
Comment (slavapestov) [inline include/swift/AST/DiagnosticsSema.def]: Line up the " with the column after ( from the previous line

**Review by slavapestov — COMMENTED** [review]
Looks reasonable but @xedin should take a look also

File: lib/Sema/TypeChecker.h
Comment (xedin) [inline lib/Sema/TypeChecker.h]: Please don't commit formatting changes like this, it makes it harder to review. Actual change here is addition to `CTP_SubscriptAssignSource`...

File: test/Constraints/assignment.swift
Comment (xedin) [inline test/Constraints/assignment.swift]: @jrose-apple What do you think about updated message?

Comment (jrose-apple) [reply test/Constraints/assignment.swift]: I'm not sure if we use the phrasing "assign to a subscript" anywhere else, but I like that it's still pretty concise, so 👍. (In particular, "subscript with element type 'X'" would be more correct but I'm not sure it's worth it.)

Comment (xedin) [reply test/Constraints/assignment.swift]: Thanks!

**Review by xedin — APPROVED** [review]
LGTM! Thank you! Once the nits @slavapestov and me mentioned this should be ready for merge.

**Comment by owenv** [comment]
Thanks for reviewing! The formatting issues should be fixed now, looks like I need to pay a little more attention to my clang-format usage.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by owenv** [comment]
Hi @xedin , do you mind merging this when you have time if  everything looks good now? Thanks!

**Comment by xedin** [comment]
Ah yes, sorry! Let's run CI again just in case and I'm going to merge.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
Done, thanks again, @owenv !

**Pull request merged**
